javalib/src/main/scala/java/security/Timestamp.scala

```diff
@@ -1,7 +1,70 @@
 package java.security
 
+// Ported from Harmony
+
 import java.util.Date
 import java.security.cert.CertPath
 
-final class Timestamp(timestamp: Date, signerCertPath: CertPath)
-    extends Serializable
+import scala.collection.mutable.StringBuilder
+
+/**
+ * Constructs a new instance of {@code Timestamp} with the specified
+ * {@code timestamp} and the given certificate path.
+ *
+ * @param timestamp date and time.
+ * @param signerCertPath the certificate path.
+ * @throws NullPointerException
+ *            if {@code timestamp} is {@code null} or if {@code signerCertPath} is {@code null}.
+ */
+final case class Timestamp(timestamp: Date, signerCertPath: CertPath)
```

Comment:
Why make it a `case class`? Is it to get `equals` for free? It's a plain `class` in Java, and we stick to the original implementation as closely as possible.

Reply:
In `new Timestamp(someDate, someSignerCertPath)`, the constructor is supposed to clone `someDate`. `Date` is mutable, and modifying `someDate` after constructing a `Timestamp` with it is not supposed to modify the `Date` held by the `Timestamp`.

Reply:
I improved it using an overrided constructor:

```scala

final class Timestamp private (private val signerCertPath: CertPath, private val timestamp: Date)
    extends Serializable {

  /**
    * Constructs a new instance of {@code Timestamp} with the specified
    * {@code timestamp} and the given certificate path.
    *
    * @param timestamp date and time.
    * @param signerCertPath the certificate path.
    * @throws NullPointerException either if [[timestamp]] or [[signerCertPath]] are `null`
    */
  @throws[NullPointerException]
  def this(timestamp: Date, signerCertPath: CertPath): Timestamp = {

    if (timestamp == null) {
      // Messages.getString("security.0F")) //$NON-NLS-1$
      throw new NullPointerException("timestamp cannot be null")
    }
    if (signerCertPath == null) {
      // Messages.getString("security.10") //$NON-NLS-1$
      throw new NullPointerException("signerCertPath cannot be null")
    }
    new Timestamp(
      timestamp = new Date(timestamp.getTime()),
      signerCertPath = signerCertPath
    )
  }
[...]
```

Reply:
Exactly, thanks!